### PR TITLE
Start writing loss Class and functions

### PR DESCRIPTION
**Pull request by PaulWask**

### CODE REVIEW

File: Car_Localizer/Car_Localizer/Base_loss_model/loss_model_base.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+
+class LossBase(nn.Module):
+    def __init__(self):
+        super(LossBase, self).__init__()
```
Comment (Siwwyy) [inline Car_Localizer/Car_Localizer/Base_loss_model/loss_model_base.py]: since python 3.8 you do need to pass explicitly a class name to `super()` method
super().__init__()

File: Car_Localizer/Car_Localizer/Base_loss_model/loss_model_base.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+        self.Alpha = nn.Parameter(torch.ones(1), requires_grad=True)
+        self.Beta = nn.Parameter(torch.zeros(1), requires_grad=True)
+        self.Sigma = nn.Parameter(torch.ones(1), requires_grad=True)
+
+    def my_bce_with_logits_loss(x, y):
+        loss = -1.0 * (y * F.logsigmoid(x) + (1 - y) * torch.log(1 - torch.sigmoid(x)))
```
Comment (Siwwyy) [inline Car_Localizer/Car_Localizer/Base_loss_model/loss_model_base.py]: please refer to my comments according to explicit types e.g.,
`x:TensorType` and function return type
if you deal with floating points, please use 1.0 instead of 1, which will be treated implicitly as integer

File: Car_Localizer/Car_Localizer/Base_loss_model/loss_model_base.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+        loss = loss.mean()
+        return loss
+
+    def my_bce_with_logits_loss_stable(x, y):
+        max_val = (-x).clamp_min_(0)
```
Comment (Siwwyy) [inline Car_Localizer/Car_Localizer/Base_loss_model/loss_model_base.py]: clamp_min_? 
look at `torch.clamp/clip` method documentation

File: Car_Localizer/Car_Localizer/Base_loss_model/loss_model_base.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+        loss = (
+            (1 - y) * x
+            + max_val
+            + torch.log(torch.exp(-max_val) + torch.exp(-x - max_val))
+        )
+        loss = loss.mean()
```
Comment (Siwwyy) [inline Car_Localizer/Car_Localizer/Base_loss_model/loss_model_base.py]: just simply `return loss.mean()`

**Review by Siwwyy — CHANGES_REQUESTED** [review]

File: Car_Localizer/Car_Localizer/Base_loss_model/loss_model_first.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+import torch
+import torch.nn as nn
+import numpy as np
+
+
+class Loss_First(Loss_Base):
+    def __init__(
+        self,
+        name: str = "Loss_First",
+        input_shape: ShapeType = (1, 3, 64, 64),
```
Comment (Siwwyy) [inline Car_Localizer/Car_Localizer/Base_loss_model/loss_model_first.py]: input shape and num classes are not needed in both Base and Loss First (It should be called `Loss_MSE`)

File: Car_Localizer/Car_Localizer/Base_loss_model/loss_model_base.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+        input_shape: ShapeType = (1, 3, 64, 64),
+        num_classes: int = 2,
+    ):
+        super().__init__()
+        self.name = name
+        self.input_shape = input_shape
+        self.num_classes = num_classes
+
+    @abstractmethod
+    def forward(self, x: TensorType = None) -> TensorType:
```
Comment (Siwwyy) [inline Car_Localizer/Car_Localizer/Base_loss_model/loss_model_base.py]: your base class should provide the same declaration
```
def forward(self, x: TensorType = None, y_pred: TensorType = None) -> TensorType:
```

File: Car_Localizer/Car_Localizer/Base_loss_model/loss_model_base.py [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+from abc import ABCMeta, abstractmethod
+from Config.config import TensorType, ShapeType
+
+
+class Loss_Base(torch.nn.Module, metaclass=ABCMeta):
+    def __init__(
+        self,
+        name: str = "Loss_Base",
+        input_shape: ShapeType = (1, 3, 64, 64),
```
Comment (Siwwyy) [inline Car_Localizer/Car_Localizer/Base_loss_model/loss_model_base.py]: get rid of that

File: Car_Localizer/Car_Localizer/Base_loss_model/loss_model_base.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+        num_classes: int = 2,
```
Comment (Siwwyy) [inline Car_Localizer/Car_Localizer/Base_loss_model/loss_model_base.py]: and that

File: Car_Localizer/Car_Localizer/Base_loss_model/loss_model_base.py [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+    ):
+        super().__init__()
+        self.name = name
+        self.input_shape = input_shape
```
Comment (Siwwyy) [inline Car_Localizer/Car_Localizer/Base_loss_model/loss_model_base.py]: keep only the name

File: Car_Localizer/Car_Localizer/Base_loss_model/loss_model_first.py [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+        num_classes: int = 2,
+    ):
+        super().__init__(name, input_shape, num_classes)
+
+    def forward(self, x: TensorType = None, y_pred: TensorType = None) -> TensorType:
+        output = np.square(np.substract(x, y_pred)).mean()
```
Comment (Siwwyy) [inline Car_Localizer/Car_Localizer/Base_loss_model/loss_model_first.py]: why you use numpy? as we work only in PyTorch
btw, I will not work, because to use PyTorch code with numpy, you have to use `x.to_numpy()`, but anyways that code is invalid, please use only `torch.###()` related methods

**Review by Siwwyy — CHANGES_REQUESTED** [review]

File: Car_Localizer/Car_Localizer/Base_loss_model/loss_model_base.py [hunk 11]
```diff
@@ -3,23 +3,19 @@
...
+from Config.config import TensorType
 
 
 class Loss_Base(torch.nn.Module, metaclass=ABCMeta):
     def __init__(
         self,
-        name: str = "Loss_Base",
-        input_shape: ShapeType = (1, 3, 64, 64),
-        num_classes: int = 2,
+        name: str = "Loss_MSE",
```
Comment (Siwwyy) [inline Car_Localizer/Car_Localizer/Base_loss_model/loss_model_base.py]: Loss_Base != Loss_MSE

just create another class, derive from Loss_Base and call it Loss_MSE

File: Car_Localizer/Car_Localizer/Base_loss_model/loss_model_base.py [hunk 12]
```diff
@@ -3,23 +3,19 @@
...
     ):
         super().__init__()
         self.name = name
-        self.input_shape = input_shape
-        self.num_classes = num_classes
 
     @abstractmethod
-    def forward(self, x: TensorType = None) -> TensorType:
+    def forward(self, x: TensorType = None, y_pred: TensorType = None) -> TensorType:
         assert x is None, "Input tensor X can't be None!"
```
Comment (Siwwyy) [inline Car_Localizer/Car_Localizer/Base_loss_model/loss_model_base.py:17]: add the same assert for `y_pred`

File: Car_Localizer/Car_Localizer/Base_loss_model/loss_mae.py [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+
+class Loss_MAE(Loss_Base):
+    def __init__(
+        self,
+        name: str = "Loss_MAE",
+    ):
+        super().__init__(name)
+
+    def forward(self, x: TensorType = None, y_pred: TensorType = None) -> TensorType:
+        output = abs(y_pred - x).mean()
```
Comment (Siwwyy) [inline Car_Localizer/Car_Localizer/Base_loss_model/loss_mae.py]: torch.abs(y_pred-x).mean()

**Review by Siwwyy — APPROVED** [review]
LGTM in overall, we can merge if you address my last comment
